js/data/bucket.js

```diff
@@ -337,32 +328,96 @@ function createElementBufferType(components) {
     });
 }
 
-function createPaintAttributes(bucket) {
-    var attributes = {};
-    for (var interfaceName in bucket.programInterfaces) {
-        var layerPaintAttributes = attributes[interfaceName] = {};
-
-        for (var c = 0; c < bucket.childLayers.length; c++) {
-            var childLayer = bucket.childLayers[c];
-            layerPaintAttributes[childLayer.id] = { attributes: [], uniforms: [] };
+function createDataLayers(bucket) {
+    var layers = {};
+
+    for (var dataLayerTypeName in bucket.dataLayerTypes) {
+        var dataLayerType = bucket.dataLayerTypes[dataLayerTypeName];
+        var dataTypeLayers = layers[dataLayerTypeName] = {};
+
+        for (var c = 0; c < bucket.styleLayers.length; c++) {
+            dataTypeLayers[bucket.styleLayers[c].id] = {
+                attributes: [],
+                uniforms: [],
+                defines: [],
+                vertexPragmas: {},
+                fragmentPragmas: {}
+            };
         }
 
-        var interface_ = bucket.programInterfaces[interfaceName];
-        if (!interface_.paintAttributes) continue;
-        for (var i = 0; i < interface_.paintAttributes.length; i++) {
-            var attribute = interface_.paintAttributes[i];
+        if (!dataLayerType.paintAttributes) continue;
+        for (var i = 0; i < dataLayerType.paintAttributes.length; i++) {
+            var attribute = dataLayerType.paintAttributes[i];
+            attribute.multiplier = attribute.multiplier || 1;
+
+            for (var j = 0; j < bucket.styleLayers.length; j++) {
+                var styleLayer = bucket.styleLayers[j];
+                var layer = dataTypeLayers[styleLayer.id];
+
+                var attributeType = attribute.components === 1 ? 'float' : 'vec' + attribute.components;
+                var attributeInputName = attribute.name;
+                assert(attribute.name.slice(0, 2) === 'a_');
+                var attributeInnerName = attribute.name.slice(2);
```

Comment:
We need to have two identifiers for each attribute:

 - one to pass the multiplied (i.e. floating point colors are multiplied by 255 to fit in a byte) value to the shader as an attribute or uniform, called `attributeInputName` above, prefixed by `a_`
 - one to hold the unmultiplied value and pass it to the vertex shader, called `attributeInnerName` above, not prefixed by `a_`

```js
        layoutAttributes: [{
            name: 'a_pos',
            components: 2,
            type: 'Int16'
        }],
        paintAttributes: [{
            name: 'a_color',
            components: 4,
            type: 'Uint8',
            getValue: function(layer, globalProperties, featureProperties) {
                return util.premultiply(layer.getPaintValue("circle-color", globalProperties, featureProperties));
            },
            multiplier: 255,
            paintProperty: 'circle-color',
            precision: 'lowp'
        }, {
            name: 'a_radius',
            components: 1,
            type: 'Uint16',
            isLayerConstant: false,
            getValue: function(layer, globalProperties, featureProperties) {
                return [layer.getPaintValue("circle-radius", globalProperties, featureProperties)];
            },
            multiplier: 10,
            paintProperty: 'circle-radius',
            precision: 'mediump'
        }]
```

I went down the road of declaring `CircleBucket`'s `paintAttributes` without the `a_` prefix and decided against it because

 - it is inconsistent with the `layoutAttributes` declarations
 - it requires more mutation of the `attribute` objects within `createPaintAttributes` (i.e. we need to append the `a_` to the name within the attribute object because that's the identifier of the actual shader attribute)

What is your concern with this `slice` operation?